### PR TITLE
feat(address-service): DOMA-8493 Search strategies were added

### PR DESCRIPTION
**Pull request by AleX83Xpert**
There are address search strategies were introduced.

TODO
- [x] tests

### Before (213ms)
image

### After (53ms)
image

### CODE REVIEW

File: apps/address-service/domains/common/utils/services/search/strategies/index.js [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+const { StrategyEachItemToPlugins } = require('./StrategyEachItemToPlugins')
+
+/**
+ * @param req
+ * @param keystone
+ * @param plugins
+ * @return {AbstractBulkSearchStrategy}
+ */
+function createBulkSearchStrategy (req, keystone, plugins) {
+    const strategy = String(get(req, ['body', 'strategy'], BULK_SEARCH_STRATEGY_EACH_ITEM_TO_PLUGINS))
```
[the remark this answers is not in the document]
Comment (AleX83Xpert) [reply apps/address-service/domains/common/utils/services/search/strategies/index.js:18]: I'll do it later, after some time in production.

File: apps/address-service/domains/common/utils/services/search/strategies/constants.js [hunk 2]
```diff
@@ -0,0 +1,26 @@
...
+const BULK_SEARCH_STRATEGIES = [
+    BULK_SEARCH_STRATEGY_EACH_ITEM_TO_PLUGINS,
+    BULK_SEARCH_STRATEGY_ALL_ITEMS_TO_EACH_PLUGIN,
+]
+
+const SEARCH_ERROR_COMMON = 'SEARCH_ERROR_COMMON'
+const SEARCH_ERROR_NO_PLUGINS = 'SEARCH_ERROR_NO_PLUGINS'
+const SEARCH_ERROR_NOT_FOUND = 'SEARCH_ERROR_NOT_FOUND'
+
+const BULK_SEARCH_CHUNK_SIZE = 5
```
[the remark this answers is not in the document]
Comment (AleX83Xpert) [reply apps/address-service/domains/common/utils/services/search/strategies/constants.js:16]: For now, the constant is enough.

File: apps/address-service/domains/common/utils/services/search/strategies/StrategyAllItemsToEachPlugin.js [hunk 3]
```diff
@@ -0,0 +1,78 @@
+const { chunk, get } = require('lodash')
+
+const { AddressFromStringParser } = require('@open-condo/clients/address-service-client/utils')
+
+const { AbstractBulkSearchStrategy } = require('./AbstractBulkSearchStrategy')
+const { BULK_SEARCH_CHUNK_SIZE, SEARCH_ERROR_NOT_FOUND, SEARCH_ERROR_COMMON } = require('./constants')
+
+class StrategyAllItemsToEachPlugin extends AbstractBulkSearchStrategy {
```
[the remark this answers is not in the document]
Comment (AleX83Xpert) [reply apps/address-service/domains/common/utils/services/search/strategies/StrategyAllItemsToEachPlugin.js:13]: Some docs were added

**Review by toplenboren — APPROVED** [review]
Why not choose something more memorable and associative as strategy names?

Like: All Items To First Plugin, then failed to second, etc could be **cascade** search or **waterfall** strategy

The second strategy, where we pass search string to all plugins might be called **parallel** strategy

LGTM, anyway

**Pull request merged**